### PR TITLE
bench: Fix bug in coin selection retry logic

### PR DESCRIPTION
**Pull request by practicalswift**
Fix bug in coin selection retry logic.

### CODE REVIEW

File: src/bench/coin_selection.cpp [hunk 1]
```diff
@@ -48,8 +48,10 @@ static void CoinSelection(benchmark::State& state)
         bool bnb_used;
         CoinEligibilityFilter filter_standard(1, 6, 0);
         CoinSelectionParams coin_selection_params(false, 34, 148, CFeeRate(0), 0);
-        bool success = wallet.SelectCoinsMinConf(1003 * COIN, filter_standard, vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)
-                       || wallet.SelectCoinsMinConf(1003 * COIN, filter_standard, vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used);
+        bool success = wallet.SelectCoinsMinConf(1003 * COIN, filter_standard, vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used);
+        if (!success) {
+            success = wallet.SelectCoinsMinConf(1003 * COIN, filter_standard, vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used);
```
Comment (Xekyo) [inline src/bench/coin_selection.cpp]: Perhaps it should also be clarified why the two are expected to have different results given that they have exactly the same input parameters. I assume that the first pass sets bnb_used to true, but @achow101 should be able to clarify.

**Review by murchandamus — COMMENTED** [review]
LGTM, I suggest that @achow101 also takes a look.

File: src/bench/coin_selection.cpp (hunk 1 above)
Comment (achow101) [reply src/bench/coin_selection.cpp]: This seems like a mistake. The first pass should be using BnB with its own CoinSelectionParams which signals for BnB to be used.

Comment (practicalswift) [reply src/bench/coin_selection.cpp]: @achow101 What is the correct `CoinSelectionParams` for the first pass? :-)

Comment (achow101) [reply src/bench/coin_selection.cpp]:
```
CoinSelectionParams coin_selection_params(true, 34, 148, CFeeRate(0), 0);
```

Comment (practicalswift) [reply src/bench/coin_selection.cpp]: @achow101 Thanks! Please re-review updated version :-)

**Comment by achow101** [comment]
utACK 313124067009e372a7d3726b6286d0001cd3ba78

File: src/bench/coin_selection.cpp [hunk 2]
```diff
@@ -47,9 +47,12 @@ static void CoinSelection(benchmark::State& state)
         CAmount nValueRet;
         bool bnb_used;
         CoinEligibilityFilter filter_standard(1, 6, 0);
-        CoinSelectionParams coin_selection_params(false, 34, 148, CFeeRate(0), 0);
-        bool success = wallet.SelectCoinsMinConf(1003 * COIN, filter_standard, vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)
-                       || wallet.SelectCoinsMinConf(1003 * COIN, filter_standard, vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used);
+        CoinSelectionParams coin_selection_params_pass_1(true, 34, 148, CFeeRate(0), 0);
+        bool success = wallet.SelectCoinsMinConf(1003 * COIN, filter_standard, vCoins, setCoinsRet, nValueRet, coin_selection_params_pass_1, bnb_used);
+        if (!success) {
+            CoinSelectionParams coin_selection_params_pass_2(false, 34, 148, CFeeRate(0), 0);
```
Comment (promag) [inline src/bench/coin_selection.cpp]: Looks like this is a 2 pass algorithm 😛 

Why not just `coin_selection_params.use_bnb = false;` and keep the same params?

File: src/bench/coin_selection.cpp [hunk 3]
```diff
@@ -47,9 +47,12 @@ static void CoinSelection(benchmark::State& state)
...
+            success = wallet.SelectCoinsMinConf(1003 * COIN, filter_standard, vCoins, setCoinsRet, nValueRet, coin_selection_params_pass_2, bnb_used);
+        }
```
Comment (promag) [inline src/bench/coin_selection.cpp]: nit, inside the retry block:
```cpp
    assert(!bnb_used);
}
```

**Review by promag — COMMENTED** [review]
~~utACK 3131240.~~

File: src/bench/coin_selection.cpp (hunk 2 above)
Comment (MarcoFalke) [inline src/bench/coin_selection.cpp]: sucess is never false, so this is dead code?

**Comment by MarcoFalke** [comment]
Everything other than `SelectCoinsMinConf` can be moved out of the hot loop, since we don't want to measure the performance of `addCoin`. (Looks like creating and deleting the wallet takes about 20% of the whole benchmark time)

File: src/bench/coin_selection.cpp (hunk 2 above)
Comment (promag) [reply src/bench/coin_selection.cpp]: You mean it should never be false as there is enough coins?

Comment (MarcoFalke) [reply src/bench/coin_selection.cpp]: Yes, otherwise the benchmark would currently fail, no?

Comment (achow101) [reply src/bench/coin_selection.cpp]: `success` can be false when BnB fails to find an exact match. The first pass should be using BnB (which is why it needs its own parameters with true set) which can fail. Then it falls back to not using BnB (which is the second params with false) which should never fail.

Comment (promag) [reply src/bench/coin_selection.cpp]: But given the current input it always succeeds. I agree that a different test exercising that case would be nice.

Comment (achow101) [reply src/bench/coin_selection.cpp]: Ah, indeed. This should have different test cases to trigger both algorithms.

**Pull request closed without merging**